### PR TITLE
Add ability to cancel build jobs

### PR DESCRIPTION
**Pull request by chefsalim**
This change adds the ability to cancel a build job group. This includes the addition of a new 'hab bldr job cancel' command, a new cancel API, and changes to the job srv and worker protocols to issue cancels. Currently, a job is cancellable between it's various stages (eg, validation, build, promotion, export, etc) - ie, it is not yet possible to cancel a job immediately when it is inside a studio build.

Marking this as Do-Not-Merge only so that we don't merge this large change prior to merging the Airlock related fixes.

Signed-off-by: Salim Alam <salam@chef.io>

tenor-40208835

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-jobsrv/src/data_store.rs [hunk 1]
```diff
@@ -191,6 +191,25 @@ impl DataStore {
         Ok(jobs)
     }
 
+    /// Get a list of cancel pending jobs
```
[the remark this answers is not in the document]
Comment (chefsalim) [reply components/builder-jobsrv/src/data_store.rs]: yep - updating

File: components/builder-jobsrv/src/server/handlers.rs [hunk 2]
```diff
@@ -212,12 +212,79 @@ pub fn job_group_abort(
...
+    // Set all the InProgress projects jobs to CancelPending
+    for project in group.get_projects().iter().filter(|&ref p| {
+        p.get_state() == jobsrv::JobGroupProjectState::InProgress
+    })
+    {
+        let job_id = project.get_job_id();
+        let mut req = jobsrv::JobGet::new();
+        req.set_id(job_id);
+
+        match state.datastore.get_job(&req)? {
```
[the remark this answers is not in the document]
Comment (chefsalim) [reply components/builder-jobsrv/src/server/handlers.rs]: will add a TODO to do so later

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 3]
```diff
@@ -264,6 +264,9 @@ impl WorkerMgr {
             if process_work ||
                 (&now > &(last_processed + Duration::from_millis(DEFAULT_POLL_TIMEOUT_MS)))
             {
+                if let Err(err) = self.process_cancels() {
+                    warn!("Worker-manager unable to process cancels: err {:?}", err);
```
[the remark this answers is not in the document]
Comment (chefsalim) [reply components/builder-jobsrv/src/server/worker_manager.rs]: yep - updating

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 4]
```diff
@@ -305,9 +308,70 @@ impl WorkerMgr {
...
+
+        if jobs.len() > 0 {
+            debug!("process_cancels: Found {} cancels", jobs.len());
+        }
+
+        for job in jobs {
+            let mut job = Job::new(job);
+
+            // Find the worker processing this job
+            let worker_ident = match self.workers.iter().find(
```
[the remark this answers is not in the document]
Comment (chefsalim) [reply components/builder-jobsrv/src/server/worker_manager.rs]: workers is already a hashmap (well, linkedhashmap), but it's hashed by worker ident.  yeah, I'd like to make this not be an iterative search as well - I'll add a TODO.

File: components/builder-api/src/server/handlers.rs [hunk 5]
```diff
@@ -181,6 +181,44 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
...
+            match id.parse::<u64>() {
+                Ok(g) => g,
+                Err(e) => {
+                    debug!("Error finding group. e = {:?}", e);
+                    return Ok(Response::with(status::BadRequest));
+                }
+            }
+        }
+        None => return Ok(Response::with(status::BadRequest)),
+    };
```
Comment (raskchanky) [inline components/builder-api/src/server/handlers.rs]: Not something that needs to be done in this PR, but I'd _love_ to find a way to extract this pattern into some kind of abstraction.  I feel like we have a ton of "find this param and return a 400 if it's not present", or "find this param and parse it into an int and return a 400 if it's not present" sprinkled throughout the depot/api components.  Maybe something we can make an issue for and tackle in the future.

File: components/builder-jobsrv/src/migrations/jobs.rs [hunk 6]
```diff
@@ -354,5 +354,17 @@ pub fn migrate(migrator: &mut Migrator) -> Result<()> {
                     "#,
     )?;
 
+    // Get a list of all the cancel pending jobs
+    migrator.migrate(
+        "jobsrv",
+        r#"CREATE OR REPLACE FUNCTION get_cancel_pending_jobs_v1()
+                     RETURNS SETOF jobs
+                     LANGUAGE SQL VOLATILE AS $$
```
Comment (raskchanky) [inline components/builder-jobsrv/src/migrations/jobs.rs]: This can probably be `STABLE` instead of `VOLATILE`, no?

**Comment by raskchanky** [comment]
A few minor comments, but looks great overall.

**Comment by raskchanky** [comment]
tenor-45520902

File: components/builder-api/src/server/handlers.rs (hunk 5 above)
Comment (chefsalim) [reply components/builder-api/src/server/handlers.rs]: yeah totally - this aspect of the handlers is ripe for some further cleanup.

File: components/builder-jobsrv/src/migrations/jobs.rs (hunk 6 above)
Comment (chefsalim) [reply components/builder-jobsrv/src/migrations/jobs.rs]: I think this does need to be volatile.. the set of cancel pending jobs changes, and the function does not return the correct set if marked stable.

Comment (raskchanky) [reply components/builder-jobsrv/src/migrations/jobs.rs]: Ahh, interesting.  I thought `VOLATILE` was just for insert/update/delete.  TIL.

**Comment by raskchanky** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296093197).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/296093197).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/296093197).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**